File: apps/farcaster-mobile/modules/expo-tus-client/ios/ExpoTusClientModule.swift
import ExpoModulesCore
import Foundation

public class ExpoTusClientModule: Module {
  private var activeUploads: [String: TUSClientUpload] = [:]
  
  public func definition() -> ModuleDefinition {
    Name("ExpoTusClient")
    
    Events("onProgress", "onSuccess", "onError", "onStart")

    AsyncFunction("createUpload") { (fileUri: String, uploadUrl: String, options: [String: Any]?) -> String in
      guard let fileUrl = URL(string: fileUri) else {
        let error = ExpoTusClientError.invalidFileUri
        throw GenericException([
          "error": error.localizedDescription,
          "code": error.errorCode
        ])
      }
      
      guard let serverUrl = URL(string: uploadUrl) else {
        let error = ExpoTusClientError.invalidUploadUrl
        throw GenericException([
          "error": error.localizedDescription,
          "code": error.errorCode
        ])
      }
      
      let upload = TUSClientUpload(
        fileUrl: fileUrl,
        serverUrl: serverUrl,
        options: options ?? [:],
        module: self
      )
      
      self.activeUploads[fileUri] = upload
      return fileUri
    }
    
    AsyncFunction("startUpload") { (fileUri: String) in
      guard let upload = self.activeUploads[fileUri] else {
        let error = ExpoTusClientError.uploadNotFound
        throw GenericException([
          "error": error.localizedDescription,
          "code": error.errorCode
        ])
      }
      
      do {
        try await upload.start()
      } catch let error as ExpoTusClientError {
        throw GenericException([
          "error": error.localizedDescription,
          "code": error.errorCode
        ])
      } catch {
        let error = ExpoTusClientError.unknownError(description: "Unknown Error")
        throw GenericException([
          "error": error.localizedDescription,
          "code": error.errorCode
        ])
      }
    }
    
    AsyncFunction("stopUpload") { (fileUri: String) in
      guard let upload = self.activeUploads[fileUri] else {
        let error = ExpoTusClientError.uploadNotFound
        throw GenericException([
          "error": error.localizedDescription,
          "code": error.errorCode
        ])
      }
      
      upload.stop()
      self.activeUploads.removeValue(forKey: fileUri)
    }

    AsyncFunction("cleanup") { (fileUri: String) in
      guard let upload = self.activeUploads[fileUri] else {
        let error = ExpoTusClientError.uploadNotFound
        throw GenericException([
          "error": error.localizedDescription,
          "code": error.errorCode
        ])
      }
      
      upload.cleanupStorageDirectory()
      self.activeUploads.removeValue(forKey: fileUri)
    }
  }
  
  internal func sendUploadEvent(_ eventName: String, fileUri: String, data: [String: Any] = [:]) {
    var eventData = data
    eventData["fileUri"] = fileUri
    sendEvent(eventName, eventData)
  }
}

File: apps/farcaster-mobile/modules/expo-tus-client/android/src/main/java/expo/modules/tusclient/ExpoTusClientModule.kt
package expo.modules.tusclient

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.net.URL
import java.util.concurrent.ConcurrentHashMap

class ExpoTusClientModule : Module() {
  private val activeUploads = ConcurrentHashMap<String, ExpoTusClientUpload>()

  override fun definition() = ModuleDefinition {
    Name("ExpoTusClient")

    Events("onProgress", "onSuccess", "onError", "onStart")

    AsyncFunction("createUpload") { fileUri: String, uploadUrl: String, options: Map<String, Any>? ->
      val file = File(fileUri)
      if (!file.exists()) {
        val error = ExpoTusClientError.invalidFileUri()
        throw Exception(error.message)
      }

      val serverUrl = try {
        URL(uploadUrl)
      } catch (e: Exception) {
        val error = ExpoTusClientError.invalidUploadUrl()
        throw Exception(error.message)
      }

      val upload = ExpoTusClientUpload(
        fileUrl = file,
        serverUrl = serverUrl,
        options = options ?: emptyMap<String, Any>(),
        module = this@ExpoTusClientModule
      )

      activeUploads[fileUri] = upload
      fileUri
    }

    AsyncFunction("startUpload") { fileUri: String ->
      val upload = activeUploads[fileUri] ?: run {
        val error = ExpoTusClientError.uploadNotFound()
        throw Exception(error.message)
      }

      try {
        upload.start()
      } catch (e: Exception) {
        val error = ExpoTusClientError.unknownError("Unknown Error")
        throw Exception(error.message)
      }
    }

    AsyncFunction("stopUpload") { fileUri: String ->
      val upload = activeUploads[fileUri] ?: run {
        val error = ExpoTusClientError.uploadNotFound()
        throw Exception(error.message)
      }

      upload.stop()
      activeUploads.remove(fileUri)
    }

    AsyncFunction("cleanup") { fileUri: String ->
      val upload = activeUploads[fileUri] ?: run {
        val error = ExpoTusClientError.uploadNotFound()
        throw Exception(error.message)
      }

      upload.cleanupStorageDirectory()
      activeUploads.remove(fileUri)
    }
  }

  internal fun sendUploadEvent(eventName: String, fileUri: String, data: Map<String, Any> = emptyMap()) {
    val eventData = data.toMutableMap()
    eventData["fileUri"] = fileUri
    sendEvent(eventName, eventData)
  }
}
